src/Symfony/Component/Messenger/Worker.php

```diff
@@ -43,12 +56,63 @@ public function run()
             });
         }
 
-        $this->receiver->receive(function (?Envelope $envelope) {
+        $this->receiver->receive(function (?Envelope $envelope, ?QueuedMessageMetadata $messageMetadata) {
             if (null === $envelope) {
                 return;
             }
 
-            $this->bus->dispatch($envelope->with(new ReceivedStamp()));
+            $this->dispatchEvent(
+                WorkerMessageHandlingEvent::class,
+                new WorkerMessageHandlingEvent($envelope, $this->receiverName)
+            );
+
+            try {
+                $this->bus->dispatch($envelope->with(new ReceivedStamp()));
+            } catch (\Throwable $throwable) {
+                $shouldRequeue = $this->shouldRequeue($throwable, $messageMetadata);
+                if ($shouldRequeue) {
+                    $this->receiver->retry($messageMetadata->getMessage());
+                } else {
+                    $this->receiver->reject($messageMetadata->getMessage());
+                }
+
+                $this->dispatchEvent(
+                    WorkerMessageFailedEvent::class,
+                    new WorkerMessageFailedEvent($envelope, $this->receiverName, $throwable, $shouldRequeue)
+                );
+
+                return;
+            }
+
+            $this->receiver->acknowledge($messageMetadata->getMessage());
+            $this->dispatchEvent(
+                WorkerMessageHandledEvent::class,
+                new WorkerMessageHandledEvent($envelope, $this->receiverName)
+            );
         });
     }
+
+    private function dispatchEvent(string $eventName, Event $event)
+    {
+        if (null === $this->eventDispatcher) {
+            return;
+        }
+
+        $this->eventDispatcher->dispatch($eventName, $event);
+    }
+
+    private function shouldRequeue(\Throwable $e, QueuedMessageMetadata $messageMetadata): bool
+    {
+        if ($e instanceof UnrecoverableMessageHandlingException) {
+            return false;
+        }
+
+        $numberOfRetries = $messageMetadata->getNumberOfTimesRetried() + 1;
+        // TODO - make retry attempts configurable.
```

Comment:
I'd suggest they should be configured via the transport DSN. Obviously you wouldn't have it anymore but you could have it as part of the `QueuedMessageMetadata` (to be moved in an envelope stamp) that could have a `getRemainingAttempts()` method or similar.

Reply:
I refactored this as well. I now use a `RetryStrategyInterface` (with a default implementation) that can be configured on a transport-by-transport basis. The logic is actually handled outside of the transport, but the user can configure on a transport-by-transport basis